bulldozer/merge.go

```diff
@@ -89,78 +117,75 @@ func MergePR(ctx context.Context, pullCtx pull.Context, client *github.Client, m
 				logger.Debug().Msg("Pull request already closed")
 				return
 			}
 
-			if pr.Mergeable == nil {
+			if mergeState.Mergeable == nil {
 				logger.Debug().Msg("Pull request mergeability not yet known")
 				continue
 			}
 
-			if !pr.GetMergeable() {
+			if !*mergeState.Mergeable {
 				logger.Debug().Msg("Pull request is not mergeable")
 				return
 			}
 
 			// Try a merge, a 405 is expected if required reviews are not satisfied
 			logger.Info().Msgf("Attempting to merge pull request with method %s", mergeOpts.MergeMethod)
-			result, _, err := client.PullRequests.Merge(ctx, pullCtx.Owner(), pullCtx.Repo(), pullCtx.Number(), commitMessage, mergeOpts)
+			sha, err := merger.Merge(ctx, pullCtx, commitMessage, mergeOpts)
 			if err != nil {
-				gerr, ok := err.(*github.ErrorResponse)
+				gerr, ok := errors.Cause(err).(*github.ErrorResponse)
 				if !ok {
-					logger.Error().Err(errors.WithStack(err)).Msg("Merge failed unexpectedly")
+					logger.Error().Err(err).Msg("Merge failed unexpectedly")
 					continue
 				}
 
 				switch gerr.Response.StatusCode {
 				case http.StatusMethodNotAllowed:
-					logger.Info().Msgf("Merge rejected due to unsatisfied condition %q", gerr.Message)
+					logger.Info().Msgf("Merge rejected due to unsatisfied condition: %q", gerr.Message)
 					return
 				case http.StatusConflict:
-					logger.Info().Msgf("Merge rejected due to being invalid %q", gerr.Message)
+					logger.Info().Msgf("Merge rejected due to being invalid: %q", gerr.Message)
 					return
 				default:
-					logger.Error().Err(errors.WithStack(err)).Msgf("Merge failed unexpectedly %q", gerr.Message)
+					logger.Error().Msgf("Merge failed with unexpected status: %d: %q", gerr.Response.StatusCode, gerr.Message)
 					continue
 				}
 			}
 
-			logger.Info().Msgf("Successfully merged pull request for sha %s with message %q", result.GetSHA(), result.GetMessage())
+			logger.Info().Msgf("Successfully merged pull request as SHA %s", sha)
 
-			// Delete ref if owner of BASE and HEAD match
-			// otherwise, its from a fork that we cannot delete
-			if pr.GetBase().GetUser().GetLogin() == pr.GetHead().GetUser().GetLogin() {
+			// if head is qualified (contains ":"), PR is from a fork and we don't have delete permission
+			if !strings.ContainsRune(head, ':') {
 				if mergeConfig.DeleteAfterMerge {
-					ref := fmt.Sprintf("refs/heads/%s", pr.Head.GetRef())
+					ref := fmt.Sprintf("refs/heads/%s", head)
 
 					// check other open PRs to make sure that nothing is trying to merge into the ref we're about to delete
-					prs, err := pull.ListOpenPullRequestsForRef(ctx, client, pullCtx.Owner(), pullCtx.Repo(), ref)
+					isTargeted, err := pullCtx.IsTargeted(ctx)
 					if err != nil {
-						logger.Error().Err(errors.WithStack(err)).Msgf("Unable to list open prs against ref %s to compare delete request", ref)
+						logger.Error().Err(err).Msgf("Unable to determine if ref %s is targeted by other open pull requests before deletion", ref)
```

Comment:
nit: Why the change away from `errors.WithStack(err)`?

Reply:
I assume that `pull.Context` implementations already wrap errors as needed, so adding a stack again is redundant. It was useful before because the code was mostly dealing with the GitHub client directly.